runtime/router/trie.go

```diff
@@ -0,0 +1,310 @@
+}
+
+type tnode struct {
+	key      string
+	value    http.Handler
+	children []*tnode
+}
+
+// NewTrie creates a new trie.
+func NewTrie() *Trie {
+	return &Trie{
+		root: &tnode{
+			key: "",
+		},
+	}
+}
+
+// Set sets the value for given path, returns error if path already set.
+// When a http.Handler is registered for a given path, a subsequent Get returns the registered
+// handler if the url passed to Get call matches the set path. Match in this context could mean either
+// equality (e.g. url is "/foo" and path is "/foo") or url matches path pattern, which has two forms:
+// - path ends with "/*", e.g. url "/foo" and "/foo/bar" both matches path "/*"
+// - path contains colon wildcard ("/:"), e.g. url "/a/b" and "/a/c" bot matches path "/a/:var"
+func (t *Trie) Set(path string, value http.Handler) error {
+	if path == "" || strings.Contains(path, "//") {
+		return errPath
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	// ignore trailing slash
+	path = strings.TrimSuffix(path, "/")
+
+	// validate "*"
+	if strings.Contains(path, "*") && !strings.HasSuffix(path, "/*") {
+		return errors.New("/* must be the last path segment")
+	}
+	if strings.Count(path, "*") > 1 {
+		return errors.New("path can not contain more than one *")
+	}
+
+	err := t.root.set(path, value, false, false)
+	if e, ok := err.(*paramMismatch); ok {
+		return fmt.Errorf("path %q has a different param key %q, it should be the same key %q as in existing path %q", path, e.actual, e.expected, e.existingPath)
+	}
+	return err
+}
+
+// Get returns the http.Handler for given path, returns error if not found.
+// It also returns the url params if given path contains any, e.g. if a handler is registered for
+// "/:foo/bar", then calling Get with path "/xyz/bar" returns a param whose key is "foo" and value is "xyz".
+func (t *Trie) Get(path string) (http.Handler, []Param, error) {
+	if path == "" || strings.Contains(path, "//") {
+		return nil, nil, errPath
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	// ignore trailing slash
+	path = strings.TrimSuffix(path, "/")
+	return t.root.get(path, false, false, false)
+}
+
+// set sets the handler for given path, creates new child node if necessary
+// lastKeyCharSlash tracks whether the previous key char is a '/', used to decide it is a pattern or not
+// when the current key char is ':'. lastPathCharSlash tracks whether the previous path char is a '/',
+// used to decide it is a pattern or not when the current path char is ':'.
+func (t *tnode) set(path string, value http.Handler, lastKeyCharSlash, lastPathCharSlash bool) error {
+	// find the longest common prefix
+	var shorterLength, i int
+	keyLength, pathLength := len(t.key), len(path)
+	if keyLength > pathLength {
+		shorterLength = pathLength
+	} else {
+		shorterLength = keyLength
+	}
+	for i < shorterLength && t.key[i] == path[i] {
+		i++
+	}
+
+	// Find the first character that differs between "path" and this node's key, if it exists.
```

Comment:
This code is quite confusing. I am guessing you are trying to have not just O(n) on the length of `key` but that you only process each character of `key` and `path` once and allocate no additional memory other than the indices. 

I think it may be much more readable if you split the string into path components and processed the equality of path components in sequence. Splitting the string into path components will roughly double the total memory used to register routes, but it's probably a better trade-off because routes are only registered once within the application.

Reply:
The trie does not necessarily split node on path segments, so splitting the path into segments and comparing equality doesn't work.

Register `/ab/c`
trie: `[/ab/c]`

Register `/a`
trie: 
```
[/a]
  |
[b/c]
```